scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -696,101 +671,54 @@ static s32 pick_idle_cpu(struct task_struct *p,
 		 * If the task can still run on the previously used CPU and
 		 * it's a full-idle core, keep using it.
 		 */
-		if (bpf_cpumask_test_cpu(prev_cpu, cast_mask(p_mask)) &&
-		    bpf_cpumask_test_cpu(prev_cpu, idle_smtmask) &&
-		    scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
+		if (bpf_cpumask_test_cpu(prev_cpu, cast_mask(cache_mask[cidx_p])) &&
+			bpf_cpumask_test_cpu(prev_cpu, idle_smtmask) &&
+			scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
 			cpu = prev_cpu;
 			goto out_put_cpumask;
 		}
 
 		/*
-		 * Search for any full-idle CPU in the primary domain that
-		 * shares the same L2 cache.
-		 */
-		if (l2_mask) {
-			cpu = bpf_cpumask_any_and_distribute(cast_mask(l2_mask), idle_smtmask);
-			if (bpf_cpumask_test_cpu(cpu, online_cpumask) &&
-			    scx_bpf_test_and_clear_cpu_idle(cpu))
-				goto out_put_cpumask;
-		}
-
-		/*
-		 * Search for any full-idle CPU in the primary domain that
-		 * shares the same L3 cache.
-		 */
-		if (l3_mask) {
-			cpu = bpf_cpumask_any_and_distribute(cast_mask(l3_mask), idle_smtmask);
-			if (bpf_cpumask_test_cpu(cpu, online_cpumask) &&
-			    scx_bpf_test_and_clear_cpu_idle(cpu))
-				goto out_put_cpumask;
-		}
-
-		/*
-		 * When considering the preferred domain (first idle CPU
-		 * selection pass) try to stay on the same LLC.
+		 * Search for any full-idle CPU that shares the same L2/L3 cache,
+		 * and then for any full-idle core, in the primary domain.
 		 */
-		if (do_preferred) {
-			cpu = -ENOENT;
-			goto out_put_cpumask;
-		}
-
-		/*
-		 * Search for any other full-idle core in the primary domain.
-		 */
-		cpu = bpf_cpumask_any_and_distribute(cast_mask(p_mask), idle_smtmask);
-		if (bpf_cpumask_test_cpu(cpu, online_cpumask) &&
-		    scx_bpf_test_and_clear_cpu_idle(cpu))
-			goto out_put_cpumask;
+#define TRY_IDLE_MASK(layer, idle_mask_type) do { \
```

Comment:
Adding `#define` inside the code makes it a little bit hard to read, maybe we can move these to the top, outside of the function definition.

Reply:
The code itself is in the best spot (as it's supposed to be used there, and never more), but you are absolutely right that the macro looks nasty.
pre-processor directives can be indented, so maybe it can be indented like the other code around to reside better.
If there's any better idea, please let me know, so I can get rid of it.